cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,384 @@
+
+        Returns:
+            A copy of this operation wrapping the provided circuit.
+
+        Raises:
+            NotImplementedError: The new circuit contains measurements, but
+                this operation has repetitions.
+        """
+        if protocols.is_measurement(new_circuit) and self._repetitions != 1:
+            raise NotImplementedError('Measurements cannot be added to looped circuits.')
+        new_op = self.copy()
+        new_op._circuit = new_circuit.freeze()
+        return new_op
+
+    def repeat(
+        self,
+        repetitions: int,
+        modulus: int = 0,
+        allow_invert: bool = False,
+        validate_modulus: bool = False,
+    ) -> 'CircuitOperation':
+        """Returns a copy of this operation repeated 'repetitions' times.
+
+        Args:
+            repetitions: Number of times this operation should repeat. This
+                is multiplied with any pre-existing repetitions.
+            modulus: If this is specified, the final repetition count will be
+                mapped to the range [0, modulus), or (-modulus/2, modulus/2]
+                if `allow_invert` is True.
+            allow_invert: For use with `modulus`. If this is True, repetitions
+                will map to the range (-modulus/2, modulus/2]. This is useful
+                for minimizing the total number of operations in a circuit.
+            validate_modulus: For use with `modulus`. If this is True, the
+                function will raise an error if the provided modulus does not
+                match the actual modulus of this operation.
+
+        Returns:
+            A copy of this operation repeated 'repetitions' times.
+
+        Raises:
+            TypeError: `repetitions` is not an integer value.
+            NotImplementedError: The operation contains measurements and
+                cannot have repetitions.
+            ValueError: `validate_modulus` was set to True, but raising this
+                operation to `modulus` does not produce the identity.
+        """
+        if not isinstance(repetitions, int):
+            raise TypeError('Only integer repetitions are allowed.')
+        if protocols.is_measurement(self.circuit):
+            raise NotImplementedError('Loops over measurements are not supported.')
+        new_op = self.copy()
+        new_op._repetitions *= repetitions
+        if modulus:
+            if validate_modulus and not linalg.allclose_up_to_global_phase(
+                protocols.unitary(self.circuit * modulus), np.eye(np.prod(self.circuit.qid_shape()))
+            ):
+                raise ValueError('Raising the circuit to "modulus" must produce the identity.')
+            new_op._repetitions %= modulus
+            if allow_invert and new_op._repetitions > modulus // 2:
+                new_op._repetitions -= modulus
+        return new_op
+
+    def __pow__(self, power: int) -> 'CircuitOperation':
+        return self.repeat(power)
```

Comment:
I don't think we can support consistent behavior for this in the long term due to other planned features. One such feature is a "loop iterator" param that assumes user-specified values for each repetition; there isn't a clear way to handle a non-integer power in such cases.

Reply:
To be clear, I'm not suggesting non-integer powers in general, only if you want to raise an already-iterated circuit to a non-integer power such that the end result is still an integer number of repetitions. This seems like it would still be compatible with loop iterators as you suggest, but I don't know exactly what you have in mind.

Reply:
In a pre-PR version of this change, loop iterators looked something like this:
```python
def repeat(self, {...}, loop_param: sympy.Symbol, loop_values: List[float]):
  # when decomposed, "loop_param" will be assigned each value in "loop_values" in turn
```
This would cause issues (or at least have unclear behavior) with something like:
```python
circuit4 = circuit.repeat(4, loop_param=sympy.Symbol('x'), loop_values: [1, 3, 6, 10])
circuit2 = circuit4.repeat(0.5)  # what values should "loop_param" be assigned?
```

Reply:
Opting to leave fractional repeats for future work.